### ISSUE TITLE
Debugging/profiling: source file cannot be found: 'sources/graal:sources/jdk/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java'

### ISSUE THREAD

**Issue by Karm** [graalvm/mandrel#103]
Hello,

it seems to me there is a source file missing in the Mandrel distro.
Source:
```java
//23 lines of comments stripped away
public class Main {
    public static void main(String[] args) {
       if (args.length<1) {
           System.out.println("Hello, Sailor!");
       } else {
           System.out.println("Hello, "+args[0]);
       }
    }
}
```
Mandrel:
```shell
$ export JAVA_HOME=~/workspaceRH/mandrel-packaging/mandrelJDK/;export GRAALVM_HOME=${JAVA_HOME};export PATH=${JAVA_HOME}/bin:${PATH}
```
Where `mandrelJDK` is a build of mandrel/mandrel/20.1 HEAD 287bbb3 using https://github.com/graalvm/mandrel-packaging master HEAD 62bd8c7 and [AdoptOpenJDK 11.0.8 GA](https://github.com/AdoptOpenJDK/openjdk11-upstream-binaries/releases/tag/jdk-11.0.8%2B10).

```shell
$ javac Main.java &&  native-image -H:GenerateDebugInfo=1 Main
$ valgrind --tool=callgrind ./main
$ kcachegrind ./callgrind.out.22052
```
Browsing the call stack...

![KCacheGrind](https://user-images.githubusercontent.com/691097/89893610-f51da100-dbd8-11ea-86eb-b09d44f3d608.png)

When I manually added the path to my original repo, `/home/karm/workspaceRH/graal/graal/compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/`, KCacheGrind found and displayed the source just fine:

![KCacheGrind-OK](https://user-images.githubusercontent.com/691097/89893662-08307100-dbd9-11ea-8e10-3049f7c80311.png)

Not sure whether we do something about it here or in https://github.com/mandrel/mandrel-packaging. I can clone/move the bug as needed. FYI @zakkak 

(Why KCacheGrind?! It used to be my weapon of choice for profiling httpd issues... This problem is not specific to the tool.)

**Comment by zakkak** [comment]
@Karm I think I am better suited to debug this, since the issue seems to be related with the comp_dir attribute that I added to debug info (assigned it to myself).

**Comment by zakkak** [comment]
The issue seems to be in `valgrind` which takes the comp_dir and appends the file name to it resulting in a path like `sources/graal:sources/jdk/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java` while the path should be `sources/graal/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java`. 
`valgrind` should instead chose the correct compilation directory from the comp_dir attribute (which may contain more than one directories in the form `path1:path2:path3`).

**Comment by Karm** [comment]
@zakkak I don't think you are right.  JDK9StringSubstitutions.java is physically not present in my /tmp/sources. The other files that valgrind finds without issue are present.

**Comment by adinn** [comment]
@zakkak it is perhaps worth investigating what is in the src.zip of the jdk11u that is being used here. My 20.2-b0.2 labs jdk11u release includes JDK9StringSubstitutions.java in the appropriate location

     -rw-r--r--      2983  29-Jun-2020  08:24:26  jdk.internal.vm.compiler/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java

So, if it is in the src.zip of the vanilla jdk11u tree it it ought to be found and copied over when the source tree is being populated.

**Comment by zakkak** [comment]
@Karm indeed `sources/graal/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java` was a left over from an older compilation.

@adinn I see no `JDK9StringSubstitutions.java` in the sources of OpenJDK 11.0.9+2
```
unzip -l /opt/jvms/openjdk-11.0.9_2/lib/src.zip | grep -i JDK9String
```

but it's present in:
```
unzip -l /home/zakkak/code/mandrel-packaging/mandrel-20.1.0.1.Alpha2.plus/lib/jvmci/graal.src.zip | grep -i JDK9String
     3137  08-10-2020 18:03   org/graalvm/compiler/replacements/JDK9StringSubstitutions.java
```

**Comment by adinn** [comment]
@zakkak yeah well it really ought to have been included in the jdk sources for jdk11u since it is used in the implementation.

I know it is in lib/jvmci/graal.src.zip in Mandeel but it really ought to be found as a jdk class in the jvmci module rather than as a Graal source. The JDK source cache is configured to field lookups for classes in the org.graalvm.compiler package. If it has a source root for that package then the Graal source cache will not get asked to do lookups.

**Comment by zakkak** [comment]
> @zakkak yeah well it really ought to have been included in the jdk sources for jdk11u since it is used in the implementation.

@jerboaa sounds like a packaging issue, can you please have a look?

**Comment by zakkak** [comment]
I was able to reproduce this with latest graal as well.

It looks like we have two separate issues here:
1. The lack of jvmci sources in src.zip of OpenJDK (I will open a new issue to track this)
2. Valgrind creating wrong file paths (which I suggest we keep track of in this issue)

As shown here: 
![image](https://user-images.githubusercontent.com/1435395/90020066-fc67ac00-dcb7-11ea-8d5e-17e9182e59fe.png)

KCacheGrind fails to show the source code for `CEntryPointSnippets.java` despite it being present in the sources directory:
```
$ find sources -name "CEntryPointSnippets.java"                       
sources/graal/com/oracle/svm/core/graal/snippets/CEntryPointSnippets.java
```

To my understanding the reason it fails is that valgrind is creating wrong file paths (by ignoring the fact that comp_dir can contain a colon separated list of directories):
```
$ grep CEntryPointSnippets callgrind.out.95187 
cfi=(156) sources/graal:sources/jdk/com/oracle/svm/core/graal/snippets/CEntryPointSnippets.java
```

while it should be:
```
cfi=(156) sources/graal/com/oracle/svm/core/graal/snippets/CEntryPointSnippets.java
```

**Comment by zakkak** [comment]
Apparently I have been abusing `comp_dir` (with the blessings of gdb). According to the dwarf spec:

> A DW_AT_comp_dir attribute whose value is a null-terminated string containing the current working directory of the compilation command that produced this compilation unit in whatever form makes sense for the host system.
>
>  _The suggested form for the value of the DW_AT_comp_dir attribute on UNIX systems is “hostname:pathname”. If no hostname is available, the suggested form is “:pathname”._ 

There is no mention about supporting multiple directories in `comp_dir` as a colon separated list.

**Comment by jerboaa** [comment]
@zakkak Are you sure this is an OpenJDK packaging issue?

> I was able to reproduce this with latest graal as well.
> 
> It looks like we have two separate issues here:
> 
>     1. The lack of jvmci sources in src.zip of OpenJDK (I will open a new issue to track this)

Hmm, I'm not sure it should be in `src.zip` of OpenJDK. The sources of OpenJDK 11 itself don't have any `JDK9StringSubstitutions.java` file:

In the OpenJDK 11u dev tree:
```
$ find src/jdk.internal.vm.compiler -name JDK9StringSubstitutions.java
```

In the mandrel source tree it is indeed found:
```
$ find compiler/ -name JDK9StringSubstitutions.java
compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java
```

This smells to me like an issue of us using `--upgrade-module-path /path/to/openjdk11/lib/svm/bin/../../jvmci/graal.jar`

**Comment by jerboaa** [comment]
As a result the debug info lookup gets confused. Trying to find it in the JDK's src.zip file while it's *exclusively* in graal.src.zip

**Comment by zakkak** [comment]
@jerboaa you are right, I was just doing a similar search, thanks for confirming. I will open a new issue to track this down.

**Comment by adinn** [comment]
So, the problem here is that packages are directed to a specific source cache by the source cache manager using a list of prefixes. The source cache manager thinks the JDK source cache owns org.graalvm.compiler on jdk11. So, if there are subpackages that need to be fielded by the Graal source cache manager then we need an exclusion for that in the lookup.

There is also another wrinkle here. package org.graalvm.compiler belongs to the jvmci compiler module in jdk11. So, if Graal is adding classes to that package it has to sneak them into the module. Now, we can still look up the sources via a src.zip that is managed by the Graal source cache but the organization of the sources in the src.zip ought to be as expected for other sources i.e. no module prefix in the dir path and noone of the other intervening dirs that have been munged into the JDK src.zip. If that is not the case then the Graal source cache will need to understand about this more complex layout organization as the JDK cache does.

**Comment by zakkak** [comment]
@adinn let's take this ^^  to #108

**Comment by adinn** [comment]
@jerboaa By the way JDK9StringSubstitutions.java is found  in the jdk dev tree

    [adinn@sputstik jdk]$ find src -name JDK9StringSubstitutions.java
    src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java
    [adinn@sputstik jdk]$ 

The file is identical modulo a trivial difference in the file header.

    diff .../JDK9StringSubstitutions.java ...//JDK9StringSubstitutions.java 
    
    7c7,9
    <  * published by the Free Software Foundation.
    ---
    >  * published by the Free Software Foundation.  Oracle designates this
    >  * particular file as subject to the "Classpath" exception as provided
    >  * by Oracle in the LICENSE file that accompanied this code.
    23,24d24
    < 
    < 

I guess it is not needed in the jdk11u code because it is only relevant for SVM. I wonder if there are other sources for which this is the case?

**Comment by jerboaa** [comment]
> @jerboaa By the way JDK9StringSubstitutions.java is found in the jdk dev tree
> 
> ```
> [adinn@sputstik jdk]$ find src -name JDK9StringSubstitutions.java
> src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/JDK9StringSubstitutions.java
> [adinn@sputstik jdk]$ 
> ```
> 
> The file is identical modulo a trivial difference in the file header.
> 
> ```
> diff .../JDK9StringSubstitutions.java ...//JDK9StringSubstitutions.java 
> 
> 7c7,9
> <  * published by the Free Software Foundation.
> ---
> >  * published by the Free Software Foundation.  Oracle designates this
> >  * particular file as subject to the "Classpath" exception as provided
> >  * by Oracle in the LICENSE file that accompanied this code.
> 23,24d24
> < 
> < 
> ```
> 
> I guess it is not needed in the jdk11u code because it is only relevant for SVM. I wonder if there are other sources for which this is the case?

@adinn I'm not surprised. The file got introduced with [JDK-8206992](https://bugs.openjdk.java.net/browse/JDK-8206992) in JDK 12. So this is a case of OpenJDK 11u being at a lower graal level . a.k.a `jdk.internal.vm.compiler` code.

**Comment by zakkak** [comment]
@adinn @jerboaa please use #108 for this part of the issue.

**Comment by zakkak** [comment]
Closed via #109
